astriaorg/astria
# feat(bridge-withdrawer): PoC astria-bridge-withdrawer implementation

## Summary
implement `ethereum` and `submitter` modules for astria-bridge-withdrawer.

## Background
required to implement rollup bridge withdrawals.

## Changes

ethereum:
- implement `AstriaWithdrawer.sol` contract, put inside `ethereum/` dir as a foundry project
- implement `ethereum` module which contains a `Watcher` that watches the contract for `Withdrawal` events
- implement event to action conversion
- the events are sent to a `Batcher` which batches all events by block number and sends a batch of actions to a sequencer handler 

sequencer: 
- submit txs to sequencer when a batch of actions is received

## Testing
unit tests

manual testing:
1. run sequencer+cometbft as normal (`just run` + `just run-cometbft` in astria-sequencer)
2. run anvil as normal (`anvil` in a terminal)
3.  init the bridge account: 
```sh
export SEQUENCER_PRIVATE_KEY=2bd806c97f0e00af1a1fc3328fa763a9269723c8db8fac4f93af71db186d6e90
./target/debug/astria-cli sequencer init-bridge-account --sequencer-url=http://localhost:26657 --rollup-name=astria --sequencer.chain-id=astria
```
4. `just copy-env` in `astria-bridge-withdrawer`, put the private key from above into a file, update .env to use that 
5. go to `astria-bridge-withdrawer/ethereum`, `cp local.env.example .env`
6. deploy the withdrawer contract:
```sh
source .env
forge create --rpc-url $RPC_URL --private-key $PRIVATE_KEY src/AstriaWithdrawer.sol:AstriaWithdrawer
```
7. run this twice, makes 2 withdrawals (forces anvil to make 2 blocks) note: the withdrawer will only submit the withdraw for the first tx, as it's waiting for the next block before submitting the second and anvil doesn't auto-build blocks by default
```sh
forge script script/AstriaWithdrawer.s.sol:AstriaWithdrawerScript \
   --rpc-url $RPC_URL --broadcast --sig "withdrawToSequencer()" -vvvv
```
8.  in `astria-bridge-withdrawer`, update `ASTRIA_BRIDGE_WITHDRAWER_ETHEREUM_CONTRACT_ADDRESS` to be the contract address deployed in step 6.
9. `just run` and you should see a withdrawal being submitted:
```sh
2024-05-24T19:36:22.952210Z DEBUG astria_bridge_withdrawer::withdrawer::submitter: fetched latest nonce nonce=1
2024-05-24T19:36:23.003380Z DEBUG astria_bridge_withdrawer::withdrawer::submitter: signed transaction tx_hash=faad567d4f71ba20bdd9a1f679c4c3875c4b172c96d236155d52eb7ca02cd389
2024-05-24T19:36:23.003742Z DEBUG submit_tx: astria_bridge_withdrawer::withdrawer::submitter: submitting signed transaction to sequencer nonce=1 transaction.hash=faad567d4f71ba20bdd9a1f679c4c3875c4b172c96d236155d52eb7ca02cd389
2024-05-24T19:36:24.049943Z  INFO astria_bridge_withdrawer::withdrawer::submitter: withdraw batch successfully executed. sequencer.block=881 sequencer.tx_hash=FAAD567D4F71BA20BDD9A1F679C4C3875C4B172C96D236155D52EB7CA02CD389 rollup.height=2
```

## Related Issues

 #913

closes #1109
